anndata/_io/merge.py

```diff
@@ -0,0 +1,647 @@
+        idx = 0
+        for arr in _gen_slice_to_append(
+            arrays, axis=axis, reindexers=reindexers, fill_value=fill_value
+        ):
+            added_size = arr.shape[axis]
+            if axis == 0:
+                out_array[idx : idx + added_size, :] = arr
+            else:
+                out_array[:, idx : idx + added_size] = arr
+            idx += added_size
+        output_group[out_path].attrs.update(
+            {"encoding-type": "array", "encoding-version": "0.2.0"}
+        )
+
+    elif isinstance(init_elem, ZarrArray):
+        import dask.array as da
+
+        darrays = (da.from_zarr(a) for a in arrays)
+        res = da.concatenate(
+            _gen_elem_to_append(
+                darrays,
+                axis=axis,
+                reindexers=reindexers,
+                fill_value=fill_value,
+            ),
+            axis=axis,
+        )
+        write_elem(output_group, out_path, res)
+        output_group[out_path].attrs.update(
+            {"encoding-type": "array", "encoding-version": "0.2.0"}
+        )
+
+
+def write_concat_sequence(
+    groups: Sequence[Union[ZarrGroup, H5Group]],
+    output_group,
+    out_path,
+    axis=0,
+    index=None,
+    reindexers=None,
+    fill_value=None,
+    join="inner",
+):
+    """
+    array, dataframe, csc_matrix, csc_matrix
+    """
+    arrays: Union[pd.DataFrame, SparseDataset, H5Array, ZarrArray] = [
+        read_group(g) for g in groups
+    ]
+    if any(isinstance(a, pd.DataFrame) for a in arrays):
+        if reindexers is None:
+            if join == "inner":
+                reindexers = gen_reindexers_df_inner(arrays, axis=axis)
+            else:
+                raise NotImplementedError("Cannot reindex dataframes with outer join.")
+        if not all(
+            isinstance(a, pd.DataFrame) or a is MissingVal or 0 in a.shape
+            for a in arrays
+        ):
+            raise NotImplementedError(
+                "Cannot concatenate a dataframe with other array types."
+            )
+        df = concat_arrays(
+            arrays=arrays,
+            reindexers=reindexers,
+            axis=axis,
+            index=index,
+            fill_value=fill_value,
+        )
+        write_elem(output_group, out_path, df)
+    # If all are compatible sparse matrices
+    elif all(
+        isinstance(a, (pd.DataFrame, SparseDataset, H5Array, ZarrArray)) for a in arrays
+    ):
+        write_concat_arrays(
+            arrays, output_group, out_path, axis, reindexers, fill_value, join
+        )
+    else:
+        raise NotImplementedError(
+            f"Concatenation of these types is not yet implemented: {[get_encoding_type(g) for g in groups] } with axis={axis}."
+        )
+
+
+def _get_group_from_hdf5(store, *args, **kwargs) -> H5Group:
+    import h5py
+
+    return h5py.File(store, *args, **kwargs)
+
+
+@singledispatch
+def _get_group(store, *args, **kwargs) -> Union[ZarrGroup, H5Group]:
```

Comment:
I rename this. Something like`"as_group"` to highlight that it's does pretty heavy coersion, and is quite dynamic.